### DISCUSSION TITLE
package name normalization for git-sourced packages

### DISCUSSION THREAD

**Discussion by cameronraysmith** [python-poetry/poetry#8781] [Q&A]
Is it possible to apply package name normalization to packages installed from git?

If I use 

```toml
HTSeq = {version = "2.0.4", markers = "sys_platform == 'linux' and platform_machine == 'x86_64'", optional = true}
```

I get

<details><summary>normalized htseq poetry.lock record</summary>
<p>

```toml
[[package]]
name = "htseq"
version = "2.0.4"
description = "A framework to process and analyze data from high-throughput sequencing (HTS) assays"
optional = false
python-versions = "*"
files = [
    {file = "HTSeq-2.0.4-cp310-cp310-macosx_10_9_x86_64.whl", hash = "sha256:ab32a9e8ff054b5faaca9be9cfbeb4b60777b53568fc05491153396cef1efdb0"},
    {file = "HTSeq-2.0.4-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl", hash = "sha256:b7d1b2835619c63fd09c7e5d7e06d4cda8aae5b520a0f345101a28b9f25ab2e0"},
    {file = "HTSeq-2.0.4-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl", hash = "sha256:506ac5a7ae6d770dc2749bfb9f382469ec61f3ad2840778deeeb04325e28ca26"},
    {file = "HTSeq-2.0.4-cp37-cp37m-macosx_10_9_x86_64.whl", hash = "sha256:ad6abdd738d5d5230d81c69c4dd9e6341cb9d3a95872cd61e91d8bbc9c71c3a4"},
    {file = "HTSeq-2.0.4-cp37-cp37m-manylinux_2_17_x86_64.manylinux2014_x86_64.whl", hash = "sha256:e74f9d2af92a1f614e6fd51c73c61fcfb205532ac8c41f4085ae21bd9e5dea19"},
    {file = "HTSeq-2.0.4-cp38-cp38-macosx_10_9_x86_64.whl", hash = "sha256:aa708e70be837293dd97bc259d86d4ae6a8b6b52af54635c6df3553e6ca42271"},
    {file = "HTSeq-2.0.4-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl", hash = "sha256:ff398e8aeeadab013f7e1b36243b97d09bc0775ee54bfda8b6ebfb91c49c9c3f"},
    {file = "HTSeq-2.0.4-cp39-cp39-macosx_10_9_x86_64.whl", hash = "sha256:ab85c30ba0ffb738ecd42f8abec59f1d719662d554db73bf63ce618180e5235e"},
    {file = "HTSeq-2.0.4-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl", hash = "sha256:dfe50aee213f9aab2051280e5ca47397a1bce98255726ffc1f9f65f7185e89d6"},
    {file = "HTSeq-2.0.4.tar.gz", hash = "sha256:5510d855617d61efff2287af4cdc60a73e94bd735e44c15b57f189ea0634e6e0"},
]

[package.dependencies]
numpy = "*"
pysam = "*"

[package.extras]
htseq-qa = ["matplotlib (>=1.4)"]
test = ["matplotlib (>=1.4)", "pandas (>=1.1.0)", "pytest (>=6.2.5)", "scipy (>=1.5.0)"]
```

</p>
</details> 

whereas if I pin to a commit on the upstream source

```toml
HTSeq = {git = "https://github.com/htseq/htseq.git", rev="d312d39187362c8c9664fa8d127ee0fb24733152", optional = true}
```

I get

<details><summary>non-normalized HTSeq poetry.lock record</summary>
<p>

```toml
[[package]]
name = "HTSeq"
version = "2.0.4"
description = "A framework to process and analyze data from high-throughput sequencing (HTS) assays"
optional = false
python-versions = "*"
files = []
develop = false

[package.dependencies]
numpy = "*"
pysam = "*"

[package.extras]
htseq-qa = ["matplotlib (>=1.4)"]
test = ["matplotlib (>=1.4)", "pandas (>=1.1.0)", "pytest (>=6.2.5)", "scipy (>=1.5.0)"]

[package.source]
type = "git"
url = "https://github.com/htseq/htseq.git"
reference = "d312d39187362c8c9664fa8d127ee0fb24733152"
resolved_reference = "d312d39187362c8c9664fa8d127ee0fb24733152"
```

</p>
</details> 

I believe this contributes to an issue where, for example,

```toml
HTSeq = [
  { markers = "sys_platform == 'linux' and platform_machine == 'x86_64'", version = "2.0.4", optional = true},
  { markers = "sys_platform == 'darwin'", git = "https://github.com/htseq/htseq.git", rev="d312d39187362c8c9664fa8d127ee0fb24733152", optional = true}
]
```

places the two records from above side-by-side in poetry.lock and package installation does not respect the markers. Perhaps this is because each record in poetry.lock only ends up being associated with one marker, but then it's impossible to achieve the intended exclusivity.
The same happens with `htseq = [...]`.
I am sure I could fork and rename the package, but that seems quite heavy-handed.

For what it's worth, this also makes it challenging to provide build inputs to the git-installed package with poetry2nix, which still seems to normalize the git-installed package name. For example, [this](https://github.com/pinellolab/DNA-Diffusion/blob/cf466b9059045db01b6f311f23b15030b7cf9080/flake.nix#L65) fails to inject the specified dependency into the build environment when using the git-installed package, even if I declare an alias in nix.

**Comment by radoering** [comment]
If it's only cosmetical, I'd say we should do it anyway for consistence if it doesn't cause any problems.

If it's really causing a bug then we should of course fix it.

**Reply by cameronraysmith** [reply]
I guess it is causing a bug with respect to providing marker-based alternatives, unless that would be considered another separate issue with markers.
It's also causing a problem in the interface with poetry2nix, but it's less clear that's a "bug".
In any case, fixing this issue would resolve the latter as well.

I converted this to an issue.

**Comment by cameronraysmith** [comment]
I closed #8783 because https://github.com/python-poetry/poetry/issues/8783#issuecomment-1853483540 suggests this behavior is intended or irrelevant from the perspective of a poetry cli user. When downstream tools choose to use the lock file contents as an input, it may be their responsibility to renormalize package names that have been left non-normalized in the lock file and determine if there are distinct but related records that might result from this.
